Fixed #26193 -- Fixed urlize to trim multiple trailing punctuation.

jdufresne opened the pull request:
<https://code.djangoproject.com/ticket/26193>

timgraham commented on django/utils/html.py:
```diff
@@ -17,7 +17,7 @@
 
 # Configuration for urlize() function.
-TRAILING_PUNCTUATION = ['.', ',', ':', ';', '.)', '"', '\'', '!']
+TRAILING_PUNCTUATION_RE = re.compile(r'^(.*?)([.,:;!]+)$')
```
Could you add a comment that explains the regular expression a bit? May not need to be as detailed as https://github.com/django/django/pull/6123 but that's an example of what I have in mind. It seems like some characters that were in `TRAILING_PUNCTUATION` aren't in the regex -- could you explain that to me so I understand it?

jdufresne replied:
> It seems like some characters that were in `TRAILING_PUNCTUATION` aren't in the regex -- could you explain that to me so I understand it?

I dropped the closing parenthesis, single quote, and double quote as they are handled by `WRAPPING_PUNCTUATION`. If I missed a detail as to why these need to exist twice, let me know and I'll add a test case for it.

timgraham commented on django/utils/html.py:
```diff
@@ -268,24 +268,40 @@ def unescape(text, trail):
             trail = ''
         return text, unescaped, trail
 
+    def trim_trailing_punctuation(lead, middle, trail):
```
Some docstrings for these functions would help. The variables aren't named such that their purpose is obvious to me.

timgraham commented on django/utils/html.py:
```diff
@@ -268,24 +268,40 @@ def unescape(text, trail):
+        for opening, closing in WRAPPING_PUNCTUATION:
+            if middle.startswith(opening):
+                middle = middle[len(opening):]
+                lead = lead + opening
```
or +=

timgraham commented on django/utils/html.py:
```diff
@@ -268,24 +268,40 @@ def unescape(text, trail):
-                    trail = closing + trail
+
+            # Continue trimming until middle is unchanged.
+            trim = True
```
`trimmed_something = True` would be more meaningful.

timgraham commented on django/utils/html.py:
```diff
@@ -268,24 +268,40 @@ def unescape(text, trail):
+            # Continue trimming until middle is unchanged.
+            trim = True
+            while trim:
+                lead, middle, trail, trimmed1 = trim_trailing_punctuation(lead, middle, trail)
```
`trimmed_punctuation` would be more meaningful.

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -324,6 +324,9 @@ Templates
 
 * Added the ``is`` comparison operator to the :ttag:`if` tag.
 
+* The :tfilter:`urlize` template filter now handles multiple trailing
```
I guess it doesn't hurt to mention, but personally I wouldn't consider this much of a "feature". :-)

jdufresne commented:
Updated commit based on feedback.

timgraham commented:
A few edits: http://dpaste.com/24AC984
I wonder if there's much advantage to the two separate trim functions? Seems like 1 method with two commented sections might be simpler.

jdufresne commented:
I updated with your edits. Thanks.

> I wonder if there's much advantage to the two separate trim functions? Seems like 1 method with two commented sections might be simpler.

Agreed. I like this better.

I consolidated the two functions down to one and updated the docblock to reflect this. The loop was also moved into the function so the function only needs to be called once.

timgraham commented:
merged in dec334cb66b3ee59cb82e1bb99a584aa0b9fbbd5, thanks!
